iocage connector that wraps around jail connector.

slohse opened the pull request:
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

iocage connection plugin
##### ANSIBLE VERSION

```
devel
```
##### SUMMARY

This change is a new connector for FreeBSD jails that are managed by iocage. It wraps around the existing jail connector and only translates the tag or UUID of an iocage-managed jail to the underlying actual name of the jail. It uses the `iocage` command to do the translation, so besides tags one can also use a partial UUID like with iocage itself.

_What is iocage and why does it need its own connector_? Iocage is a popular tool that simplifies creating and managing FreeBSD jails. But iocage uses UUIDs for the name property of a jail and provides its own "tag" property for legible names. So while the jails created by iocage can be managed in ansible with the existing jail connector, one would have to use the jail _name_ as remote host with the result that inventories and ansible run outputs would only contain hostnames such as `ioc-b39796cc-8f0d-11e6-8d6b-001d9263409c`.

Commit:
added iocage connector that wraps around jail connector. Uses iocage to translate iocage tags or UUIDs/partial UUIDs to the actual jail name and then uses the jail connector for actual functionality

mattclay commented:
@slohse The [tests](https://app.shippable.com/runs/584f22919559830f0009086c/1/console) are failing due to missing boilerplate in `lib/ansible/plugins/connection/iocage.py`:

```
2016-12-12 23:03:35 Run command: test/sanity/code-smell/boilerplate.sh
2016-12-12 23:03:35 
2016-12-12 23:03:35 == Missing __metaclass__ = type ==
2016-12-12 23:03:35 ./lib/ansible/plugins/connection/iocage.py
2016-12-12 23:03:35 
2016-12-12 23:03:35 == Missing from __future__ import (absolute_import, division, print_function) ==
2016-12-12 23:03:35 ./lib/ansible/plugins/connection/iocage.py
2016-12-12 23:03:35 ERROR: Command "test/sanity/code-smell/boilerplate.sh" returned exit status 2.
```

Here's an example of the needed boilerplate:

https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/__init__.py#L20

slohse committed:
added missing boilerplate

mattclay commented:
@slohse The failure was due to an unrelated CI issue which has been fixed. Tests are now passing.

slohse commented:
Excellent, thank you!

mattclay commented:
@slohse This definitely looks useful. I'm particularly interested in this as it should make it easier to add integration test coverage for `jail` to our existing CI process.

I discussed this briefly with @abadger and we think it would be more appropriate to inherit from the existing `jail` connection plugin instead of wrapping it. You'll need to make a few minor changes to the existing plugin to permit setting the jail name without changing `self._play_context.remote_addr`.

slohse committed:
rewrote iocage connector to inherit from jail connector

slohse commented:
Okay, I modified `jail` now in a way, that i accepts a custom jailname via `kwargs`. I'm not entirely convinced I found the most elegant solution, though.

mattclay commented on lib/ansible/plugins/connection/jail.py:
```diff
@@ -54,6 +56,8 @@ def __init__(self, play_context, new_stdin, *args, **kwargs):
         super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
 
         self.jail = self._play_context.remote_addr
+        if(self.modified_jailname_key in kwargs):
```
Remove the parenthesis from the conditional, as they're not needed.

slohse replied:
Whoops, my non-pythonic background shows :D

mattclay commented on lib/ansible/plugins/connection/iocage.py:
```diff
@@ -0,0 +1,68 @@
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
```
Unused import.

slohse replied:
Right you are

mattclay commented on lib/ansible/plugins/connection/iocage.py:
```diff
@@ -0,0 +1,68 @@
+
+        kwargs[Jail.modified_jailname_key] = 'ioc-{}'.format(jail_uuid)
+
+        display.vvv(u"Jail {iocjail} has been translated to {rawjail}".format(
```
Include a `host=kwargs[Jail.modified_jailname_key]` kwarg when calling `display.vvv`.

This will maintain consistency with other calls to `display.vvv` made by the `Jail` connection plugin.

slohse replied:
Wouldn't it be helpful in terms of error tracing to use the initial iocage tag (i.e. `self.ioc_jail`) instead of the already converted jail uuid for `host`?

mattclay replied:
That would be ideal, as long as you maintain the log entry you already have to indicate the mapping between the two.

You'll need to update the `Jail` plugin so there's a separate `self.jail_friendly_name` or something similar which is used for the `host` passed to `display.vvv`, so you're able to show the iocage jail name on all the messages from the underlying `Jail` plugin.

slohse replied:
Hm, I meant in that particular place in `iocage.py` only. I think trying to pass that down to `jail.py` could get confusing quickly and kind of blurs the edges of the two connectors.

mattclay replied:
It's important to have the same `host` kwarg value used for all calls to `display.vvv` for the same connection, whether that be the actual jail name or the iocage jail name.

As written currently, the `display.vvv` call above is simply lacking the `host` kwarg. It can be added without making any changes to the message displayed.

slohse replied:
Oh, okay. If it has to be the same throughout the connection, then I guess using the actual jail name is the most sensible approach

mattclay reviewed (changes requested):
This looks really good, except for a few minor issues.

Thanks for the prompt response to feedback.

slohse committed:
iocage connector cleanup

- removed unnecessary import
- removed unnecessary parentheses
- added host to display.vvv call

mattclay commented:
@slohse Thanks for the connection plugin. I'll take a look at using this to add iocage and jail to CI.

slohse commented:
Cool, thank you :)
